batchflow/plot.py

```diff
@@ -0,0 +1,1228 @@
+        return new_color
+
+    def add_colorbar(self, ax_image, width=.2, pad=None, color='black', fake=False):
+        """ Append colorbar to the image on the right. """
+        divider = axes_grid1.make_axes_locatable(ax_image.axes)
+        cax = divider.append_axes("right", size=width, pad=pad)
+
+        if fake:
+            cax.set_axis_off()
+            colorbar = None
+        else:
+            colorbar = ax_image.axes.figure.colorbar(ax_image, cax=cax)
+            colorbar.ax.yaxis.set_tick_params(color=color, labelcolor=color)
+
+        return colorbar
+
+    def add_legend(self, ax=None, mode='imshow', handles=None, label=None, color=None,
+                   alpha=1, size=10, ha=None, va=None, **kwargs):
+        """ TODO Add patches to legend. All invalid colors are filtered.
+
+        Notes to self: Rewrite doc, explain line/patches parametrization.
+        """
+        if isinstance(ax, int):
+            ax = self.axes[ax]
+
+        legend = ax.get_legend()
+        old_handles = getattr(legend, 'legendHandles', [])
+        texts = getattr(legend, 'get_texts', lambda: [])()
+        old_labels = [t._text for t in texts] # pylint: disable=protected-access
+
+        if mode in ('imshow', 'hist', 'wiggle'):
+            colors = to_list(color)
+            alphas = alpha if isinstance(alpha, list) else [alpha] * len(colors)
+            new_handles = []
+            # pylint: disable=redefined-argument-from-local
+            for color, alpha in zip(colors, alphas):
+                if is_color_like(color):
+                    patch = Patch(color=color, alpha=alpha)
+                    new_handles.append(patch)
+            new_labels = [] if label is None else to_list(label)
+        elif mode in ('curve', 'loss'):
+            new_handles = handles
+            new_labels = [line.get_label() for line in handles]
+
+        if len(new_handles) > 0 and len(new_labels) > 0:
+            kwargs['handles'] = old_handles + new_handles
+            kwargs['labels'] = old_labels + new_labels
+
+            legend = ax.legend(prop={'size': size}, **kwargs)
+
+        if ha is not None:
+            _ = [text.set_ha(ha) for text in legend.get_texts()]
+        if va is not None:
+            _ = [text.set_ha(va) for text in legend.get_texts()]
+
+    @staticmethod
+    def add_grid(ax, grid_type, x_n=None, y_n=None, zorder=0, **kwargs):
+        """ TODO """
+        if grid_type == 'minor':
+            locator = AutoMinorLocator
+        elif grid_type == 'major':
+            locator = MaxNLocator
+
+        if x_n is not None:
+            set_locator = getattr(ax.xaxis, f'set_{grid_type}_locator')
+            set_locator(locator(x_n))
+
+        if y_n is not None:
+            set_locator = getattr(ax.yaxis, f'set_{grid_type}_locator')
+            set_locator(locator(y_n))
+
+        ax.grid(which=grid_type, zorder=zorder, **kwargs)
+
+
+
+def plot_image(data, **kwargs):
+    """ Shorthand for image plotting. """
```

Comment:
I would prefer to make some effort to keep the docstring of the class in this function. In the simplest case, you can use `partial` or explicitly set the `__doc__` of this function.

In more advanced, you could remove all of the sections corresponding to other modes and keep only the relevant part of documentation